lang/funcs/core/os/exec_func.go

```diff
@@ -0,0 +1,198 @@
+	obj.init = init
+	obj.closeChan = make(chan struct{})
+	return nil
+}
+
+// Stream returns the changing values that this func has over time.
+func (obj *ExecFunc) Stream() error {
+	// Close the output chan to signal that no more values are coming.
+	defer close(obj.init.Output)
+
+	// A channel which closes when the current process exits, on its own
+	// or due to cancel(). The channel is only closed once all the pending
+	// stdout and stderr lines have been processed.
+	//
+	// The channel starts closed because no process is running yet. A new
+	// channel is created each time a new process is started. We never run
+	// more than one process at a time.
+	processedChan := make(chan struct{})
+	close(processedChan)
+
+	// Wait for the current process to exit, if any.
+	defer func() {
+		<-processedChan
+	}()
+
+	// Kill the current process, if any. A new cancel function is created
+	// each time a new process is started.
+	var ctx context.Context
+	var cancel context.CancelFunc
+	cancel = func() {}
+	defer cancel()
+
+	for {
+		select {
+		case input, more := <-obj.init.Input:
+			if !more {
+				// Wait until the current process exits and all of its
+				// stdout is sent downstream.
+				select {
+				case <-processedChan:
+					return nil
+				case <-obj.closeChan:
+					return nil
+				}
+			}
+			shellCommand := input.Struct()["shell_command"].Str()
+
+			// Kill the previous command, if any.
+			cancel()
+			<-processedChan
+
+			// Run the command, connecting it to ctx so we can kill
+			// it if needed, and to two Readers so we can read its
+			// stdout and stderr.
+			ctx, cancel = context.WithCancel(context.Background())
+			cmd := exec.CommandContext(ctx, "sh", "-c", shellCommand)
+			stdoutReader, err := cmd.StdoutPipe()
+			if err != nil {
+				return err
+			}
+			stderrReader, err := cmd.StderrPipe()
+			if err != nil {
+				return err
+			}
+			if err = cmd.Start(); err != nil {
+				return err
+			}
+
+			// We will now start several goroutines:
+			// 1. To process stdout
+			// 2. To process stderr
+			// 3. To wait for (1) and (2) to terminate and close processedChan
+			//
+			// This WaitGroup is used by (3) to wait for (1) and (2).
+			wg := &sync.WaitGroup{}
+
+			// Emit one value downstream for each line from stdout.
+			// Terminates when the process exits, on its own or due
+			// to cancel().
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+
+				stdoutScanner := bufio.NewScanner(stdoutReader)
+				for stdoutScanner.Scan() {
+					outputValue := &types.StrValue{V: stdoutScanner.Text()}
+					obj.init.Output <- outputValue
+				}
+			}()
+
+			// Log the lines from stderr, to help the user debug.
+			// Terminates when the process exits, on its own or
+			// due to cancel().
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+
+				stderrScanner := bufio.NewScanner(stderrReader)
+				for stderrScanner.Scan() {
+					obj.init.Logf("ExecFunc: \"%v\": stderr: %v\n", shellCommand, stderrScanner.Text())
```

Comment:
nit: consider changing the prefix to `exec` or (if we do the rename) to `system` to match what the user would expect it be called?